include/xtensor/xstrides.hpp

```diff
@@ -250,31 +292,25 @@ namespace xt
         };
     }
 
-    template <class offset_type, class S>
-    inline offset_type data_offset(const S&) noexcept
-    {
-        return offset_type(0);
-    }
-
-    template <class offset_type, class S, class Arg, class... Args>
-    inline offset_type data_offset(const S& strides, Arg arg, Args... args) noexcept
+    template <class offset_type, class S, class... Args>
+    inline offset_type data_offset(const S& strides, Args... args) noexcept
```

Comment:
I think keeping the definition with the extra `args` would allow to keep the recursive call to `data_offset` hen there are too many arguments.

Reply:
I think this would be the implementation then:
```
    template <class offset_type, class S>
    inline offset_type data_offset(const S&) noexcept
    {
        return offset_type(0);
    }

    template <class offset_type, class S>
    inline offset_type data_offset(const S& strides, missing_type arg) noexcept
    {
        return offset_type(0);
    }

    template <class offset_type, class S, class Arg>
    inline offset_type data_offset(const S& strides, Arg arg) noexcept
    {
        constexpr std::size_t nargs = 1;
        if (nargs == strides.size())
        {
            // Correct number of arguments: iterate
            return static_cast<offset_type>(detail::raw_data_offset<0>(strides, arg));
        }
        else if (nargs > strides.size())
        {
            // Too many arguments: drop the first
            return 0;
        }
        else
        {
            // Too few arguments: right to left scalar product
            auto view = strides.cend() - nargs;
            return static_cast<offset_type>(detail::raw_data_offset<0>(view, arg));
        }
    }

    template <class offset_type, class S, class Arg, class... Args>
    inline offset_type data_offset(const S& strides, Arg arg, Args... args) noexcept
    {
        constexpr std::size_t nargs = sizeof...(Args) + 1;
        if (nargs == strides.size())
        {
            // Correct number of arguments: iterate
            return static_cast<offset_type>(detail::raw_data_offset<0>(strides, arg, args...));
        }
        else if (nargs > strides.size())
        {
            // Too many arguments: drop the first
            return data_offset<offset_type, S>(strides, args...);
        }
        else
        {
            // Too few arguments: right to left scalar product
            auto view = detail::last_type_is_missing<Args...> ? 0 : strides.cend() - nargs;
            return static_cast<offset_type>(detail::raw_data_offset<0>(view, arg, args...));
        }
    }
```

Reply:
I think it couls be better to make `last_type_is_missing` a struct that you specialized when no argument is passed, i.e. something like:

```cpp
template <class... Args>
struct last_type_is_missing_impl
    : std::is_same<missing_type, mpl::back_t<mpl::vector<Args...>> 
{
};

template <>
struct last_type_is_missing_impl<>
    : std::false_type
{
};

template <class... Args>
 constexpr bool last_type_is_missing = last_type_is_missing_impl::value;
```